File: Binary Tree/Java/VerticalTraversal.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class NodeLocation {
    TreeNode root;
    int hd, vd;

    public NodeLocation(TreeNode root, int hd, int vd) {
        this.root = root;
        this.hd = hd;
        this.vd = vd;
    }

}

class Solution {
    public void traverse(TreeNode root, ArrayList<int[]> list, int hd, int vd) {
        if (root == null) {
            return;
        }
        int arr[] = { root.val, hd, vd };
        list.add(arr);
        traverse(root.left, list, hd - 1, vd + 1);
        traverse(root.right, list, hd + 1, vd + 1);
    }

    public List<List<Integer>> verticalTraversal(TreeNode root) {
        ArrayList<int[]> list = new ArrayList<>();
        traverse(root, list, 0, 0);
        Collections.sort(list, (a, b) -> {
            if (a[1] == b[1]) {
                if (a[2] == b[2]) {
                    return a[0] - b[0];
                } else {
                    return a[2] - b[2];
                }
            }
            return a[1] - b[1];
        });
        HashMap<Integer, ArrayList<Integer>> map = new LinkedHashMap<>();
        for (int[] i : list) {
            if (!map.containsKey(i[1])) {
                map.put(i[1], new ArrayList<>());
            }
            map.get(i[1]).add(i[0]);
        }
        List<List<Integer>> ans = new ArrayList<>();
        for (int i : map.keySet()) {
            ans.add(map.get(i));
        }
        return ans;
    }
}

public class VerticalTraversal {
    public static void main(String[] args) {
        TreeNode node = new TreeNode(1);
        node.left = new TreeNode(2);
        node.left.left = new TreeNode(4);
        node.left.right = new TreeNode(5);
        node.right = new TreeNode(3);
        node.right.left = new TreeNode(6);
        node.right.right = new TreeNode(7);
        Solution sol = new Solution();
        System.out.println(sol.verticalTraversal(node));
    }
}

File: Binary Tree/C++/VerticalTraversal.cpp
#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <algorithm>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
private:
    void verticalTraversal(TreeNode *root, vector<tuple<int, int, int>> &list, int hd, int vd)
    {
        if (root == nullptr)
        {
            return;
        }
        list.push_back({root->val, hd, vd});
        verticalTraversal(root->left, list, hd - 1, vd + 1);
        verticalTraversal(root->right, list, hd + 1, vd + 1);
    }

public:
    vector<vector<int>> verticalTraversal(TreeNode *root)
    {
        vector<tuple<int, int, int>> list;
        verticalTraversal(root, list, 0, 0);
        sort(list.begin(), list.end(), [](tuple<int, int, int> a, tuple<int, int, int> b)
             {
            if (get<1>(a) == get<1>(b)) {
				if (get<2>(a) == get<2>(b)) {
					return get<0>(a) < get<0>(b);
				} else {
					return get<2>(a) <get<2>(b);
				}
			}
			return get<1>(a) < get<1>(b); });
        map<int, vector<int>> map;
        for (tuple<int, int, int> i : list)
        {
            map[get<1>(i)].push_back(get<0>(i));
        }
        vector<vector<int>> ans;
        for (pair<int, vector<int>> i : map)
        {
            ans.push_back(i.second);
        }
        return ans;
    }
};
int main()
{
    TreeNode *node = new TreeNode(1);
    node->left = new TreeNode(2);
    node->left->left = new TreeNode(4);
    node->left->right = new TreeNode(5);
    node->right = new TreeNode(3);
    node->right->left = new TreeNode(6);
    node->right->right = new TreeNode(7);
    Solution sol;
    vector<vector<int>> ans = sol.verticalTraversal(node);
    for (vector<int> i : ans)
    {
        for (int j : i)
        {
            cout << j << " ";
        }
        cout << "\n";
    }
    return 0;
}
